File: sample-conference-videochat/sample-conference-videochat/Utils/AttachmentDownloadManager/AttachmentDownloadOperation/AttachmentDownloadOperation.m
//
//  AttachmentDownloadOperation.m
//  sample-conference-videochat
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "AttachmentDownloadOperation.h"
#import <Quickblox/Quickblox.h>

@implementation AttachmentDownloadOperation

//MARK: - Life Cycle
- (instancetype)initWithAttachmentID:(NSString *)attachmentID
                      attachmentName:(NSString *)attachmentName
                      attachmentType:(AttachmentType)attachmentType
                     progressHandler:(ProgressHandler)progressHandler
                      successHandler:(SuccessHandler)successHandler
                        errorHandler:(ErrorHandler)errorHandler {
    self = [super init];
    if (self) {
        self.attachmentID = attachmentID;
        self.attachmentName = attachmentName;
        self.attachmentType = attachmentType;
        self.progressHandler = progressHandler;
        self.successHandler = successHandler;
        self.errorHandler = errorHandler;
    }
    return self;
}

- (void)main {
    if (self.isCancelled) {
        self.state = AsyncOperationStateFinished;
        return;
    }
    
    [self downloadAttachmentWithID:self.attachmentID attachmentName:self.attachmentName attachmentType:self.attachmentType];
}

//MARK: - Internal Methods
- (void)downloadAttachmentWithID:(NSString *)ID attachmentName:(NSString *)attachmentName attachmentType:(AttachmentType)attachmentType {
    self.state = AsyncOperationStateExecuting;
    
    __weak typeof(self)weakSelf = self;
    
    [QBRequest downloadFileWithUID:ID successBlock:^(QBResponse * _Nonnull response, NSData * _Nonnull fileData) {
        
        if (attachmentType == AttachmentTypeImage && [UIImage imageWithData:fileData]) {
            UIImage *image = [UIImage imageWithData:fileData];
            weakSelf.successHandler(image, nil, ID);
        } else {
            NSString *fileName = [NSString stringWithFormat:@"%@_%@", ID, attachmentName];
            NSString *path = [NSTemporaryDirectory() stringByAppendingPathComponent:fileName];
            NSURL *fileURL = [NSURL fileURLWithPath:path];
            if ([fileData writeToURL:fileURL atomically:YES]) {
                weakSelf.successHandler(nil, fileURL, ID);
            }
            else {
                NSLog(@"failure");
            }
        }
        weakSelf.state = AsyncOperationStateFinished;
        
    } statusBlock:^(QBRequest * _Nonnull request, QBRequestStatus * _Nonnull status) {
        CGFloat progress = status.percentOfCompletion;
        weakSelf.progressHandler(progress, ID);
    } errorBlock:^(QBResponse * _Nonnull response) {
        weakSelf.errorHandler(response.error.error, ID);
        weakSelf.state = AsyncOperationStateFinished;
    }];
}

@end

File: sample-chat-swift/sample-chat-swift/Components/ChatManager/Operations/AttachmentDownloadOperation.swift
//
//  AttachmentDownloadOperation.swift
//  sample-chat-swift
//
//  Created by Injoit on 1/28/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

import Foundation
import UIKit
import AVKit

typealias ErrorHandler = (_ error: Error?, _ ID: String) -> Void
typealias SuccessHandler = (_ image: UIImage?, _ videoUrl: URL?, _ ID: String) -> Void
typealias ProgressHandler = (_ progress: CGFloat, _ ID: String) -> Void

final class AttachmentDownloadOperation: AsyncOperation {
    
    //MARK: - Properties
    private var successHandler: SuccessHandler?
    private var progressHandler: ProgressHandler?
    private var errorHandler: ErrorHandler?
    var attachmentID: String
    var attachmentName: String
    var attachmentType: AttachmentType
    
    //MARK: - Life Cycle
    public required init (attachmentID: String,
                          attachmentName: String,
                          attachmentType: AttachmentType,
                          progress: @escaping ProgressHandler,
                          success: @escaping SuccessHandler,
                          error: @escaping ErrorHandler) {
        self.attachmentID = attachmentID
        self.attachmentName = attachmentName
        self.progressHandler = progress
        self.successHandler = success
        self.errorHandler = error
        self.attachmentType = attachmentType
        
        super.init()
    }
    
    //MARK: - Overrides
    override func main() {
        guard isCancelled == false else {
            self.state = .finished
            return
        }
        self.downloadAttachmentWithID(attachmentID, attachmentName: attachmentName, attachmentType: attachmentType)
    }
    
    //MARK: - Internal Methods
    private func downloadAttachmentWithID(_ ID: String, attachmentName: String, attachmentType: AttachmentType) {
        self.state = .executing
        QBRequest.downloadFile(withUID: ID, successBlock: { [weak self] (response: QBResponse, fileData: Data)  in
            if attachmentType == .Image, let image = UIImage(data: fileData) {
                let fixImage = image.fixOrientation()
                CacheManager.shared.store(fixImage, for: ID)
                self?.successHandler?(fixImage.fixOrientation(), nil, ID)
            } else {
                let fileData = fileData as NSData
                let fileName = ID + "_" + attachmentName
                let filePath = NSTemporaryDirectory() + fileName
                let fileURL = URL(fileURLWithPath: filePath)
                if  fileData.write(to: fileURL, atomically: true) == true {
                    CacheManager.shared.getFileWith(stringUrl: fileURL.absoluteString) { result in
                        switch result {
                        case .success(let fileURl):
                            fileURl.getThumbnailImage { thumbnailImage in
                                CacheManager.shared.store(thumbnailImage, for: ID)
                                self?.successHandler?(thumbnailImage, fileURl, ID)
                            }
                        case .failure(let error):
                            debugPrint(error, "[AttachmentDownloadManager]  failure in the Cache of video")
                            self?.successHandler?(nil, nil, ID)
                        }
                    }
                } else {
                    debugPrint("[AttachmentDownloadOperation] failure")
                }
            }
            self?.state = .finished
        }, statusBlock: { [weak self] (request: QBRequest, status: QBRequestStatus?) in
            guard let status = status else {
                return
            }
            let progress = CGFloat(status.percentOfCompletion)
            self?.progressHandler?(progress, ID)
        }, errorBlock: { [weak self] (response: QBResponse) in
            self?.errorHandler?(response.error?.error, ID)
            self?.state = .finished
        })
    }
}
